src/jenkins.py

```diff
@@ -705,3 +708,186 @@ def get_agent_name(unit_name: str) -> str:
     """
     return unit_name.replace("/", "-")
+
+
+PLUGIN_NAME_GROUP = r"^([a-zA-Z0-9-_]+)"
+WHITESPACE = r"\s*"
+VERSION_GROUP = r"\((.*?)\)"
+DEPENDENCIES_GROUP = r"\[(.*?)\]"
+PLUGIN_CAPTURE = rf"{PLUGIN_NAME_GROUP}{WHITESPACE}{VERSION_GROUP}"
+PLUGIN_LINE_CAPTURE = rf"{PLUGIN_CAPTURE} => {DEPENDENCIES_GROUP}"
+
+
+def _get_plugin_name(plugin_info: str) -> str:
+    """Get plugin name given a plugin info string of format <plugin-shortname> (<plugin-version>).
+
+    Args:
+        plugin_info: Text containing plugin name and plugin version.
+
+    Raises:
+        ValidationError: if the plugin info string does not conform to expected format.
+
+    Returns:
+        The plugin shortname.
+    """
+    match = re.match(PLUGIN_CAPTURE, plugin_info)
+    if not match:
+        raise ValidationError(f"No plugin matched in: {plugin_info}")
+    return match.group(1)
+
+
+def _build_dependencies_lookup(
+    plugin_dependency_outputs: typing.Iterable[str],
+) -> dict[str, tuple[str, ...]]:
+    """Build a lookup table of plugin short name to list of dependency plugin's short names.
+
+    Args:
+        plugin_dependency_outputs: The plugin dependency output from Jenkins Groovy script.
+
+    Returns:
+        The dependency lookup table.
+    """
+    dependency_lookup: dict[str, tuple[str, ...]] = {}
+    for line in plugin_dependency_outputs:
+        match = re.match(PLUGIN_LINE_CAPTURE, line)
+        if not match:
+            continue
+        plugin, dependencies = match.group(1), match.group(3)
+        if not dependencies:
+            dependency_lookup[plugin] = ()
+        try:
+            dependency_lookup[plugin] = tuple(
+                _get_plugin_name(dependency) for dependency in dependencies.split(", ")
+            )
+        except ValidationError as exc:
+            logger.error("Invalid plugin dependency, %s", exc)
+            continue
+
+    return dependency_lookup
+
+
+def _get_allowed_plugins(
+    allowed_plugins: typing.Iterable[str],
+    dependency_lookup: typing.Mapping[str, typing.Iterable[str]],
+    seen: set[str] | None = None,
+) -> typing.Iterable[str]:
+    """Get the plugin short names of allowed plugins and their dependencies.
+
+    Args:
+        allowed_plugins: The allowed plugins short names to add to allowed plugins with their
+            dependencies.
+        dependency_lookup: The plugin dependency lookup table.
+        seen: Whether the plugin has been yielded already during recursive traversal.
+
+    Yields:
+        The allowed plugin short name.
+    """
+    # mypy doesn't understand that we can reassign the type and it cannot be None aftwards.
+    if seen is None:
+        seen: set[str] = set()  # type: ignore
+    for plugin in allowed_plugins:
+        if plugin in seen:  # type: ignore
+            continue
+        yield plugin
+        seen.add(plugin)  # type: ignore
+        try:
+            yield from _get_allowed_plugins(dependency_lookup[plugin], dependency_lookup, seen)
```

Comment:
Generally try and limit the scope of `try...except` to the minimal code where the intended error could occur, e.g., this approach catches KeyErrors in the recursive call and on `dependency_lookup[plugin]`